src/lib/hooks/useCurrencyColor.ts

```diff
@@ -0,0 +1,78 @@
+import { Currency } from '@uniswap/sdk-core'
+import { useTheme } from 'lib/theme'
+import Vibrant from 'node-vibrant/lib/bundle'
+import { useEffect, useLayoutEffect, useState } from 'react'
+
+import useCurrencyLogoURIs from './useCurrencyLogoURIs'
+
+const colors = new Map<string, string | undefined>()
+
+/**
+ * Extracts the prominent color from a token.
+ * NB: If cached, this function returns synchronously; using a callback allows sync or async returns.
+ */
+async function getColorFromLogoURIs(logoURIs: string[], cb: (color: string | undefined) => void = () => void 0) {
+  const key = logoURIs[0]
+  let color = colors.get(key)
+
+  if (!color) {
+    for (const logoURI of logoURIs) {
+      let uri = logoURI
+      if (logoURI.startsWith('http')) {
+        // Color extraction must use a CORS-compatible resource, but the resource may already be cached.
+        // Adds a dummy parameter to force a different browser resource cache entry. Without this, color extraction prevents resource caching.
+        uri += '?color'
+      }
+
+      color = await getColorFromUriPath(uri)
+      if (color) break
+    }
+  }
+
+  colors.set(key, color)
+  return cb(color)
+}
+
+async function getColorFromUriPath(uri: string): Promise<string | undefined> {
+  try {
+    const palette = await Vibrant.from(uri).getPalette()
+    return palette.Vibrant?.hex
+  } catch {}
+  return
+}
+
+export function usePrefetchCurrencyColor(token?: Currency) {
+  const theme = useTheme()
+  const logoURIs = useCurrencyLogoURIs(token)
+
+  useEffect(() => {
+    if (theme.tokenColorExtraction && token) {
+      getColorFromLogoURIs(logoURIs)
+    }
+  }, [token, logoURIs, theme.tokenColorExtraction])
+}
+
+export default function useCurrencyColor(token?: Currency) {
+  const [color, setColor] = useState<string | undefined>(undefined)
+  const theme = useTheme()
+  const logoURIs = useCurrencyLogoURIs(token)
+
+  useLayoutEffect(() => {
```

Comment:
I'm not super familiar w/ this hook so I looked up its usage. It looks like it's supposed to be used for reading rendered dom stuff. At first I thought it was appropriate here because of `getColorFromLogoURIs`, but after looking at the implementation, I'm not so sure.

I thiiink this should be a normal `useEffect`

Reply:
`useLayoutEffect` is necessary to update the color in the same frame as the token is updated. Without it, the effect will only run in the frame _after_ the token is updated, and there will be one frame rendered with incorrect colors, which is noticeable.